Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs

```diff
@@ -454,20 +610,183 @@ public override void Initialize()
+                    }
+                    else // Assume piping volume is 20% of full cylinder volume
+                        CylPipeVolumeM3 = 0.20f * CylVolumeM3; 
+                }
+
+                // Advanced brake cylinder sim needs a spring pressure defined, otherwise it can remain 0
+                if (BrakeCylinderSpringPressurePSI <= 0)
+                    BrakeCylinderSpringPressurePSI = 5.0f;
+
+                // Precalculate the relationship between air in cylinder lines and cylinder travel
+                if (CylTravelTab == null) // May have been initialized already
+                {
+                    // Assumptions from PRR characteristic curve:
+                    // pg 40: https://ia600906.us.archive.org/24/items/braketestsreport00penn/braketestsreport00penn.pdf
+                    // Brake cylinder reaches rated travel at 50 psi
+                    // Brake cylinder travel is 80% of nominal when pressure reaches the spring counter-pressure
+                    // Other assumptions:
+                    // Brake cylinder travel remains at 0 until reaching half the spring counter-pressure
+                    // Brake cylinder travel is linear w.r.t. total air in the cylinder line, not entirely accurate but greatly simplifies code
+                    float[] airPV = new float[7];
+                    float[] cylTravel = new float[7];
+
+                    float nomPSI = 50.0f;
+
+                    // Prevent interpolator error. Max cyl pressure and nominal pressure need to be slightly offset
+                    if (ReferencePressurePSI == nomPSI)
+                        nomPSI -= 5.0f;
+                    // Prevent interpolator error. Spring pressure should never be this large
+                    if (BrakeCylinderSpringPressurePSI >= nomPSI)
+                        BrakeCylinderSpringPressurePSI = nomPSI - 5.0f;
+
+                    // 0 cylinder travel with 0 air in cylinder
+                    cylTravel[0] = 0;
+                    airPV[0] = AdvancedBrakeCylinderAir(0, true);
+                    // 0 cylinder travel when pressure is insufficient to budge spring
+                    cylTravel[1] = 0;
+                    airPV[1] = AdvancedBrakeCylinderAir(BrakeCylinderSpringPressurePSI / 2.0f, true);
+                    // 80% cylinder travel when spring pressure is fully overcome
+                    cylTravel[2] = CylStrokeM * 0.8f;
+                    airPV[2] = AdvancedBrakeCylinderAir(BrakeCylinderSpringPressurePSI, true);
+                    // 100% cylinder travel at 50 psi, with linear change in travel between 50 psi and max pressure
+                    // This will work whether the max cylinder pressure is above or below 50 psi
+                    float strokePerPsi = (CylStrokeM - cylTravel[2]) / (50.0f - BrakeCylinderSpringPressurePSI);
```

Comment:
I'd like to see more information describing these constants (80%, 50psi, etc.). Are they arbitrary or based in any principle? Using constants may not be adequate for all braking systems

Reply:
This is all based on a characteristic curve established by Pennsylvania Railroad in [this document](https://ia600906.us.archive.org/24/items/braketestsreport00penn/braketestsreport00penn.pdf) (URL is also in the comments in the source code). Never found any other reference for brake cylinder travel vs pressure, so this was really the only thing to base an implementation off of. Obviously, the report was for one specific size of brake cylinder, so to adjust for other sizes I assume the relationship is the same, just scaled up or down. The one constant is that the 'nominal' cylinder travel is defined at 50 psi, which is the standard pressure used for brake cylinder measurements.

Obviously these assumptions could be overwritten with user input hypothetically, but that adds a lot of complexity (the goal here is for someone to be able to add just a few tokens to an old file) and relies on getting input data that simply seems to not exist. If brake cylinder travel vs pressure becomes a popular field of study, that can change.

Reply:
Sorry, I didn't see the URL. It's fine then